java-checks/src/main/java/org/sonar/java/checks/spring/SpringBeansShouldBeAccessibleCheck.java

```diff
@@ -0,0 +1,145 @@
+    + "Either move it to a package configured in @ComponentsScan or update your @ComponentsScan configuration.";
+
+  private static final String[] SPRING_BEAN_ANNOTATIONS = {
+    "org.springframework.stereotype.Component",
+    "org.springframework.stereotype.Service",
+    "org.springframework.stereotype.Repository",
+    "org.springframework.stereotype.Controller",
+    "org.springframework.web.bind.annotation.RestController"
+  };
+
+  private static final String COMPONENT_SCAN_ANNOTATION = "org.springframework.context.annotation.ComponentScan";
+  private static final Set<String> COMPONENT_SCAN_ARGUMENTS = new HashSet<>(Arrays.asList("basePackages", "value"));
+
+  private static final String SPRING_BOOT_APP_ANNOTATION = "org.springframework.boot.autoconfigure.SpringBootApplication";
+
+  /**
+   * The key is the class fully qualified name prefix, which includes the name of the package
+   * The value is a list of messages which are independent of Syntax Trees (to avoid keeping references to all ASTs in all files)
+   */
+  private final Map<String, List<AnalyzerMessage>> messagesPerClassPrefix = new HashMap<>();
+  private final Set<String> scannedPackages = new HashSet<>();
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return Collections.singletonList(Tree.Kind.CLASS);
+  }
+
+  @Override
+  public void endOfAnalysis() {
+    AnalyzerMessageReporter reporter = (AnalyzerMessageReporter) context;
+    messagesPerClassPrefix.entrySet().stream()
+      // support sub-packages and inner classes
+      .filter(entry -> scannedPackages.stream().noneMatch(entry.getKey()::contains))
+      .forEach(entry -> entry.getValue().forEach(reporter::reportIssue));
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    ClassTree classTree = (ClassTree) tree;
+
+    if (!hasSemantic() || classTree.simpleName() == null) {
+      return;
+    }
+
+    String classPackageName = extractClassPrefix(classTree.symbol().type().fullyQualifiedName());
+    SymbolMetadata classSymbolMetadata = classTree.symbol().metadata();
+
+    List<SymbolMetadata.AnnotationValue> componentScanValues = classSymbolMetadata.valuesForAnnotation(COMPONENT_SCAN_ANNOTATION);
+    if (componentScanValues != null) {
+      componentScanValues.forEach(this::addToScannedPackages);
+    } else if (hasAnnotation(classSymbolMetadata, SPRING_BOOT_APP_ANNOTATION)) {
+      scannedPackages.add(classPackageName);
+    } else if (hasAnnotation(classSymbolMetadata, SPRING_BEAN_ANNOTATIONS)) {
+      addMessageToMap(classPackageName, classTree.simpleName());
+    }
+  }
+
+  private void addMessageToMap(String classPackageName, IdentifierTree classNameTree) {
+    AnalyzerMessageReporter reporter = (AnalyzerMessageReporter) context;
+    AnalyzerMessage analyzerMessage = reporter.createAnalyzerMessage(this, classNameTree, String.format(MESSAGE_FORMAT, classNameTree));
```

Comment:
The message formatting should be relying on classnameTree.name  not on the toString of the tree.